tests/unit/test___init__.py

```diff
@@ -0,0 +1,113 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Unit tests for execution."""
```

Comment:
I have the feeling that you're retesting some parts which are already unit tested
I would mock the calls to imported method and only test the "business logic" of the run method given the results of mocked methods

I had the same feeling for integration tests, but since it's integration we need to test all components altogether, but here we're on unit test

Reply:
Yes I agree these tests are marginal in the value they provide, I'll write down some thoughts on why they were added and also on mocking.

Starting with mocking, there is value in limiting its use as much as possible because it can get complex quickly and readability goes down the more mocks are used. It also tends to go towards essentially re-writing the source code because the tendency is to check that certain calls did or did not happen and the order etc. The main reasons for mocking are either slow code or calls to external systems in unit tests because this increases the speed of the tests whilst still checking the business logic to a degree. Interactions with external systems should be behind an interface (as they are in this repo for discourse) and have suitable integration tests. After that, the remaining unit tests can reliably mock the external dependency whilst enforcing that the interface spec is being adhered to (which is achieved here using `spec=discourse.Discourse` which raises exceptions if the interface is being used incorrectly, such as calling a non-existent function or forgetting to pass a required variable). The penalty for running other code that is fast or doesn't interact very much with the outside world is relatively minor. If there is a bug with that code, it should cause test failures there which can then be run in isolation using the pytest `-k` option.

The reason to have a small number of unit tests looking at `run` is just to exercise the code a bit and surface any issues around passing the wrong arguments or errors with the logic in that function (e.g., empty versus non empty local index and the dictionary comprehension of reports). If there were no integration tests, it would be necessary to have a lot more unit tests and more extensive interaction checks here. The value of having at least some unit tests is so that obvious mistakes or errors get surfaced more quickly as integration tests take a few minutes to run.

Does that make sense? Is there anything else in particular that would be worth mocking here? Are there any tests that don't provide much value? Are there any others that would be worth adding?

Reply:
The reasoning behind my remark is that scope of unit test should be as limited as possible to avoid test failing in cascade. Let's say a bug is introduced on sub methods, it could result of MANY tests failing in the `run` method while this issue is not lying there at all.
That make finding the root cause more complex, as it could be misleading. That has been part of my conventions for long that unit test should limit their scope to the class they're testing as long as all the external methods are unit tested themselves. We've always looked at which exceptions and returned values are possible, if they influence the caller feedback and mock their behavior in different test scenarios.
This has drawbacks, a change of signature or a new possibly raised exception will require maintenance on these unit tests as they won't test this scenario.
Other convention I've always had also is, we don't test the framework, for example let's say we use an ORM that relies on a mapper to map query results to entities, we don't test that the mapper does its job properly (unless we did some override).

I'm fine with keeping them as they are, if we make sure we stay coherent over the charms over this way of reasoning for unit tests